MNE/utils/sphere.cpp

```diff
@@ -0,0 +1,338 @@
+//    % http://www.mathworks.com/matlabcentral/fileexchange/22643
+
+//    % Input:
+//    % X: n x 3 matrix of cartesian data
+//    % Outputs:
+//    % Center: Center of sphere
+//    % Radius: Radius of sphere
+//    % Author:
+//    % Alan Jennings, University of Dayton
+
+//    A=[mean(X(:,1).*(X(:,1)-mean(X(:,1)))), ...
+//        2*mean(X(:,1).*(X(:,2)-mean(X(:,2)))), ...
+//        2*mean(X(:,1).*(X(:,3)-mean(X(:,3)))); ...
+//        0, ...
+//        mean(X(:,2).*(X(:,2)-mean(X(:,2)))), ...
+//        2*mean(X(:,2).*(X(:,3)-mean(X(:,3)))); ...
+//        0, ...
+//        0, ...
+//        mean(X(:,3).*(X(:,3)-mean(X(:,3))))];
+//    A=A+A.';
+//    B=[mean((X(:,1).^2+X(:,2).^2+X(:,3).^2).*(X(:,1)-mean(X(:,1))));...
+//        mean((X(:,1).^2+X(:,2).^2+X(:,3).^2).*(X(:,2)-mean(X(:,2))));...
+//        mean((X(:,1).^2+X(:,2).^2+X(:,3).^2).*(X(:,3)-mean(X(:,3))))];
+//    Center=(A\B).';
+//    Radius=sqrt(mean(sum([X(:,1)-Center(1),X(:,2)-Center(2),X(:,3)-Center(3)].^2,2)));
+
+Sphere Sphere::fit_sphere(const MatrixX3f& points)
+{
+    const VectorXf& x = points.col(0);
+    const VectorXf& y = points.col(1);
+    const VectorXf& z = points.col(2);
+
+    VectorXf point_means = points.colwise().mean();
+
+    VectorXf x_mean_free = x.array() - point_means(0);
+    VectorXf y_mean_free = y.array() - point_means(1);
+    VectorXf z_mean_free = z.array() - point_means(2);
+
+    Matrix3f A;
+    A << (x.cwiseProduct(x_mean_free)).mean(), 2*(x.cwiseProduct(y_mean_free)).mean(), 2*(x.cwiseProduct(z_mean_free)).mean(),
+                                            0,   (y.cwiseProduct(y_mean_free)).mean(), 2*(y.cwiseProduct(z_mean_free)).mean(),
+                                            0,                                      0,   (z.cwiseProduct(z_mean_free)).mean();
+
+    Matrix3f A_T = A.transpose();
+    A += A_T;
+
+    Vector3f b;
+    VectorXf sq_sum = x.array().pow(2)+y.array().pow(2)+z.array().pow(2);
+    b << (sq_sum.cwiseProduct(x_mean_free)).mean(),
+         (sq_sum.cwiseProduct(y_mean_free)).mean(),
+         (sq_sum.cwiseProduct(z_mean_free)).mean();
+
+    Vector3f center = A.ldlt().solve(b);
+
+    MatrixX3f tmp(points.rows(),3);
+    tmp.col(0) = x.array() - center(0);
+    tmp.col(1) = y.array() - center(1);
+    tmp.col(2) = z.array() - center(2);
+
+    float r = sqrt(tmp.array().pow(2).rowwise().sum().mean());
+
+    return Sphere(center, r);
+}
+
+
+//*************************************************************************************************************
+
+Sphere Sphere::fit_sphere_simplex(const MatrixX3f& points, double simplex_size)
+{
+    VectorXf center;
+    float R;
+    fit_sphere_to_points( points, simplex_size, center, R);
+
+    return Sphere(center, R);
+}
+
+
+//*************************************************************************************************************
+//ToDo Replace LayoutMaker::fit_sphere_to_points
+bool Sphere::fit_sphere_to_points ( const MatrixXf &rr,
```

Comment:
input/output variabels underneath eacht other? is this consistent with our coding convention?